### PR TITLE
Add a hack to our docs to avoid cache

### LINKED ISSUE #910: Use unique identifier to avoid caching
I enabled the RTD PR builder for this project, however CSS and JS get cached at the CDN now, making PR review of built documentation not very helpful -- these assets will get cached and any updates to CSS/JS on the PR won't show.

We can avoid this by using a query string to avoid the caching and would have a few things to decide on:

* What string to use? We surface `{{ commit }}` for most builds, we could conditionally add that on the end of js/css requests: `{% if commit %}?{{ commit }}{% endif %}`. Is there a better string?
* Is there a more generalized solution to this? Sphinx extension that does this for us?
* Do we want this on just development/PR builds? On the base theme `layout.html` and all projects that use our theme? Perhaps start by making this a configuration variable and work towards making it the default?

My vote is to use `commit`, and use a configuration variable to enable the addition to URLs.

### PR DESCRIPTION
**Pull request by agjohnson**
It was hard to use the PR builder to test style changes, due to the CDN.
This helps avoid most of the caching for URLs in HTML. Webpack uses file
hashing during reference, so everything else should avoid caching.

Closes #910 -- we can discuss caching of PRs or a more general solution
somewhere else.

### CODE REVIEW

**Comment by ericholscher** [comment]
@agjohnson what about the CDN made this hard? It should clear the CDN cache on each build.

**Comment by ericholscher** [comment]
Ahh! The .build domain doesn't have the fancy CDN, I'll see about fixing that.

File: docs/conf.py [hunk 1]
```diff
@@ -118,3 +119,24 @@ def setup(app):
...
+    def event_context_no_cache(app, pagename, templatename, context, doctree):
+        pathto = context['pathto']
+
+        def pathto_no_cache(*args, **kwargs):
+            uri = pathto(*args, **kwargs)
+            if re_cache_files.match(uri):
+                postfix = str(int(time.time()))
+                if 'commit' in context:
+                    postfix = context['commit']
+                uri = "{uri}?{postfix}".format(uri=uri, postfix=postfix)
```
[the remark this answers is not in the document]
Comment (agjohnson) [reply docs/conf.py:137]: Yeah, that was my original thought as well, but not sure the best way to do that. I'd be :+1: on this as a feature (or perhaps even a sphinx addition/extension?)

**Comment by agjohnson** [comment]
Yeah, I still seemed to get cache hits on CSS when I tested, but I didn't poke around much.

**Comment by ericholscher** [comment]
@agjohnson I've turned the CDN off for the .build domain for now -- we don't need it.

File: docs/conf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidfischer) [reply docs/conf.py:137]: It probably needs to be an extension. Editing people's HTML and moving their static files after the fact would probably be fine in 99%+ of cases but that 1%....

**Comment by agjohnson** [comment]
Cool, makes sense. This PR might still be handy for avoiding browser caching for now. I would eventually like a general solution to the issue though.

**Comment by jessetan** [comment]
Is this for the theme assets (JS and CSS) only, or also for the generated docs? If only for JS and CSS, we could use subresource integrity for the hash

**Comment by agjohnson** [comment]
> Is this for the theme assets (JS and CSS) only, or also for the generated docs? If only for JS and CSS, we could use subresource integrity for the hash

This is just for theme assets and images directly referenced from the theme templates -- webpack already uses a hashed filename to reference files included by URL in sass sources.

This won't affect generated docs that use the theme, though I would be :+1: on figuring out how to add that feature. So far, this PR is just used to avoid caching of assets for PR builds on RTD for the theme.

Subresource integrity could definitely be one way we could potentially solve this for our theme. If we wanted to solve this at a broader level, patching the `pathto` function to include some hash would cover all themes without modification.

I'll open an issue to discuss adding this at the theme template level so that all projects using the theme would be guaranteed to have fresh assets.

**Comment by agjohnson** [comment]
Pushing this on to next release, CDN cache changes will resolve some of the issues with caching. It would be nice to be able to simply refresh and not have to worry about browser cache on PR builds while iterating on CSS. In case the CDN cache changes aren't enough, I'll keep this one around.

**Comment by agjohnson** [comment]
I don't think we hit issues with this since, but also we've been tuning our QA process for this package and the PR builds aren't a great option due to our test matrix depth. I'll close this out, it can always be resurrected if we need.

**Pull request closed without merging**